# Neural Network

In [0]:
import numpy as np
import matplotlib.pyplot as plt

## Load Dataset

In [0]:
from keras.datasets import mnist

Using TensorFlow backend.


In [0]:
((X_train, y_train), (X_test, y_test)) = mnist.load_data()

print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

11493376/11490434 [==============================] - 1s 0us/step
(60000, 28, 28) (60000,)
(10000, 28, 28) (10000,)


In [0]:
## Processing

In [0]:
X_train = X_train.reshape(60000, 28*28)
X_test = X_test.reshape(10000, 28*28)

print(X_train.shape, X_test.shape)

(60000, 784) (10000, 784)


In [0]:
from keras.utils import to_categorical

y_train_hot = to_categorical(y_train)
y_test_hot = to_categorical(y_test)

print(y_train_hot.shape, y_test_hot.shape)
print('Label = {0}'.format(y_train[0:3]))
y_train_hot[0:3]

(60000, 10) (10000, 10)
Label = [5 0 4]


array([[0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.]], dtype=float32)

## Modeling

### Multi-layer Neutral Network

### Multi_model_1

*w1=+-1, size=(784, 1000)<br>
*w2=+-1, size=(1000, 10)<br>
*lr=0.00001<br>
*epochs=100<br>

In [0]:
from keras.models import Sequential
from keras.layers import Dense
from keras.initializers import RandomUniform
from keras.optimizers import SGD

In [0]:
model = Sequential()
model.add(Dense(units=1000,
               kernel_initializer=RandomUniform(minval=-1, maxval=1),
               activation='sigmoid',
               input_dim=28*28))

model.add(Dense(units=10,
               kernel_initializer=RandomUniform(minval=-1, maxval=1),
               activation='sigmoid'))

In [0]:
# loss=Cross Entropy
# optimizer=Stochastic Gradient Descent(SGD)
model.compile(loss='categorical_crossentropy',
             optimizer=SGD(lr=0.00001),
             metrics=['accuracy'])

In [0]:
import time
def nowtime(past):
  now = time.time()
  print(now)
  print('period[second] : {}'.format(now-past))
  return now

pasttime=0

In [0]:
pasttime=time.time()
model.fit(X_train, y_train_hot, epochs=1000)
print(nowtime(pasttime))

Epoch 1/100
60000/60000 [==============================] - 17s 283us/step - loss: 9.3458 - acc: 0.0871
Epoch 2/100
60000/60000 [==============================] - 16s 274us/step - loss: 9.1995 - acc: 0.0868
Epoch 3/100
60000/60000 [==============================] - 17s 280us/step - loss: 9.0605 - acc: 0.0861
Epoch 4/100
60000/60000 [==============================] - 17s 279us/step - loss: 8.9293 - acc: 0.0859
Epoch 5/100
60000/60000 [==============================] - 17s 281us/step - loss: 8.8046 - acc: 0.0853
Epoch 6/100
60000/60000 [==============================] - 17s 287us/step - loss: 8.6856 - acc: 0.0849
Epoch 7/100
60000/60000 [==============================] - 18s 296us/step - loss: 8.5727 - acc: 0.0850
Epoch 8/100
60000/60000 [==============================] - 17s 287us/step - loss: 8.4655 - acc: 0.0850
Epoch 9/100
60000/60000 [==============================] - 17s 290us/step - loss: 8.3638 - acc: 0.0845
Epoch 10/100
60000/60000 [==============================] - 17s 287us/ste

In [0]:
import pandas as pd
y_predict = model.predict(X_test)
y_predict = np.argmax(y_predict, axis=1)

accuracy = (y_predict == y_test).mean()

print('Accuracy = {0:.5f}'.format(accuracy))
pd.DataFrame({'y(actual)':y_test, 'y(predict)':y_predict}).head(15)

Accuracy = 0.11350


y(actual)  y(predict)
0           7           3
1           2           1
2           1           3
3           0           3
4           4           8
5           1           3
6           4           5
7           9           2
8           5           1
9           9           3
10          0           0
11          6           0
12          9           3
13          0           0
14          1           0

In [0]:
# Single-layer Neural Network

In [0]:
from keras.models import Sequential
from keras.layers import Dense
from keras.initializers import RandomUniform
from keras.optimizers import SGD

In [0]:
model_s =Sequential()
model_s.add(Dense(units=10,
                 kernel_initializer=RandomUniform(minval=-1, maxval=1),
                 activation='sigmoid',
                 input_dim=28*28))

# loss = Cross Entropy
# optimizer = Stochastic Gradient Descent(SGD)
model_s.compile(loss='categorical crossentropy',
               optimizer=SGD(;r=0.00001),
               metrics=['accuracy'])

In [0]:
model.fit(X_train, y_train_hot, epochs=100)

In [0]:
import pandas as pd
y_predict = model_s.predict(X_test)
y_predict = np.argmax(y_predict, axis=1)

accuracy = (y_predict == y_test).mean()

print('Accuracy = {0:.5f}'.format(accuracy))
pd.DataFrame({'y(actual)':y_test, 'y(predict)':y_predict}).head(15)

### => Multi_model1 결과

*w1=+-1, size=(784, 1000)<br>
*w2=+-1, size=(1000, 10)<br>
*lr=0.00001<br>
*epochs=100<br>

*time = 27m<br>

*Accuracy = 0.11350<br>

### Multi_model2

*w1=+-1, size=(784, 1000)<br>
*w2=+-1, size=(1000, 10)<br>
*lr=0.00001<br>
*epochs=300<br>

In [0]:
model = Sequential()
model.add(Dense(units=1000,
               kernel_initializer=RandomUniform(minval=-1, maxval=+1),
               activation='sigmoid',
               input_dim=28*28))
model.add(Dense(units=10,
               kernel_initializer=RandomUniform(minval=-1, maxval=+1),
               activation='sigmoid'))

# loss=Cross Entropy
# optimizer=Stochastic Gradient Descent(SGD)
model.compile(loss='categorical_crossentropy', optimizer=SGD(lr=0.0001), 
              metrics=['accuracy'])

In [0]:
import time
def nowtime(past):
  now = time.time()
  print(now)
  print('period[second] : {}'.format(now-past))
  return now

pasttime=0

In [0]:
pasttime=time.time()
model.fit(X_train, y_train_hot, epochs=300)
print(nowtime(pasttime))

Epoch 1/300
60000/60000 [==============================] - 16s 271us/step - loss: 4.7205 - acc: 0.0966
Epoch 2/300
60000/60000 [==============================] - 16s 263us/step - loss: 3.5717 - acc: 0.0994
Epoch 3/300
60000/60000 [==============================] - 16s 258us/step - loss: 3.0127 - acc: 0.1038
Epoch 4/300
60000/60000 [==============================] - 15s 258us/step - loss: 2.7434 - acc: 0.1093
Epoch 5/300
60000/60000 [==============================] - 16s 264us/step - loss: 2.5943 - acc: 0.1155
Epoch 6/300
60000/60000 [==============================] - 16s 265us/step - loss: 2.5042 - acc: 0.1211
Epoch 7/300
60000/60000 [==============================] - 16s 267us/step - loss: 2.4450 - acc: 0.1270
Epoch 8/300
60000/60000 [==============================] - 16s 263us/step - loss: 2.4039 - acc: 0.1320
Epoch 9/300
60000/60000 [==============================] - 16s 263us/step - loss: 2.3743 - acc: 0.1358
Epoch 10/300
60000/60000 [==============================] - 16s 266us/ste

In [0]:
import pandas as pd

y_predict = model.predict(X_test)
y_predict = np.argmax(y_predict, axis=1)

accuracy = (y_predict == y_test).mean()

print('Accuracy = {0:.5f}'.format(accuracy))
pd.DataFrame({'y(actual)':y_test, 'y(predict)':y_predict}).head(15)

Accuracy = 0.16850


y(actual)  y(predict)
0           7           4
1           2           5
2           1           4
3           0           0
4           4           7
5           1           1
6           4           4
7           9           4
8           5           4
9           9           4
10          0           4
11          6           4
12          9           4
13          0           4
14          1           5

### => Multi_model2 결과

*w1=+-1, size=(784, 1000)<br>
*w2=+-1, size=(1000, 10)<br>
*lr=0.00001<br>
*epochs=300<br>

*time = 79m<br>

*Accuracy = 0.11350<br>